### ISSUE TITLE
Replicating en-fr UNMT with a smaller emb_dim

### ISSUE THREAD

**Issue by magician-david** [microsoft/MASS#4]
I have to use a smaller emb_dim of 512 because my GPU Mem is 12.8G. As a result, the hidden_dim of Transformer would be 4 * 512 = 2048. However, the results seem to be much worse than yours.

My command of pre-training is:
`python train.py --exp_name unsupMT_enfr --data_path './data/processed/en-fr/' --lgs 'en-fr' --bt_steps 'en-fr-en,fr-en-fr' --mass_steps 'en,fr' --lambda_bt '0:0,10:0' --encoder_only false --emb_dim 512 --n_layers 6 --n_heads 8 --dropout '0.1' --attention_dropout '0.1' --gelu_activation true --tokens_per_batch 3000 --optimizer 'adam_inverse_sqrt,beta1=0.9,beta2=0.98,lr=0.0001' --epoch_size 200000 --max_epoch 100 --eval_bleu true --word_mass '0.5' --min_len 5 --exp_id "ht6lz6ziu1"`

My command of fine-tuning is:
`python train.py --exp_name unsupMT_enfr --data_path './data/processed/en-fr/' --lgs 'en-fr' --bt_steps 'en-fr-en,fr-en-fr' --encoder_only false --emb_dim 512 --n_layers 6 --n_heads 8 --dropout '0.1' --attention_dropout '0.1' --gelu_activation true --tokens_per_batch 2000 --batch_size 32 --bptt 256 --optimizer 'adam_inverse_sqrt,beta1=0.9,beta2=0.98,lr=0.0001' --epoch_size 200000 --max_epoch 30 --eval_bleu true --reload_model 'mass_enfr_512.pth,mass_enfr_512.pth' --exp_id "w83s8z4nkx"`

Only the emb_dim is changed, but I get the following result:

pre-train.log:
> 2 days, 12:15:39 - __log__:{"epoch": 99, "valid_fr-en_mt_ppl": 249.58064788484583, "valid_fr-en_mt_acc": 23.941589780961678, "valid_fr-en_mt_bleu": 2.28, "valid_en-fr_mt_ppl": 209.35770915056546, "valid_en-fr_mt_acc": 23.806986740933585, "valid_en-fr_mt_bleu": 2.6, "test_fr-en_mt_ppl": 216.85211556087498, "test_fr-en_mt_acc": 24.96583143507973, "test_fr-en_mt_bleu": 2.53, "test_en-fr_mt_ppl": 180.5567727326117, "test_en-fr_mt_acc": 24.483515007722094, "test_en-fr_mt_bleu": 2.71}

fine-tune.log:
> 2 days, 0:07:54 - __log__:{"epoch": 29, "valid_fr-en_mt_ppl": 206.69015785463145, "valid_fr-en_mt_acc": 40.49940187275702, "valid_fr-en_mt_bleu": 10.23, "valid_en-fr_mt_ppl": 156.52663532211892, "valid_en-fr_mt_acc": 40.216746382514984, "valid_en-fr_mt_bleu": 10.53, "test_fr-en_mt_ppl": 145.12240020417357, "test_fr-en_mt_acc": 43.62819539357125, "test_fr-en_mt_bleu": 11.71, "test_en-fr_mt_ppl": 109.9388737280739, "test_en-fr_mt_acc": 43.34333102043557, "test_en-fr_mt_bleu": 12.17}

Do I need to change other hyperparams? Is an emb_dim of 1024 necessary?

**Comment by StillKeepTry** [comment]
@magician-david how many gpus do you use?

**Comment by StillKeepTry** [comment]
@magician-david and how many data do you use?

**Comment by magician-david** [comment]
> @magician-david how many gpus do you use?

One. 

I met a problem with multi-gpu. Here is my command:

`MODEL=mass_enfr_1024.pth

export NGPU=2; CUDA_VISIBLE_DEVICES="2,3" python -m torch.distributed.launch --nproc_per_node=$NGPU train.py \
  --exp_name unsupMT_enfr                              \
  --data_path ./data/processed/en-fr/                  \
  --lgs 'en-fr'                                        \
  --bt_steps 'en-fr-en,fr-en-fr'                       \
  --encoder_only false                                 \
  --emb_dim 1024                                       \
  --n_layers 6                                         \
  --n_heads 8                                          \
  --dropout 0.1                                        \
  --attention_dropout 0.1                              \
  --gelu_activation true                               \
  --tokens_per_batch 2000                              \
  --batch_size 32	                                     \
  --bptt 256                                           \
  --optimizer adam_inverse_sqrt,beta1=0.9,beta2=0.98,lr=0.0001 \
  --epoch_size 200000                                  \
  --max_epoch 30                                       \
  --eval_bleu true                                     \
  --reload_model "$MODEL,$MODEL"                       \
`

**Comment by StillKeepTry** [comment]
@magician-david can you set `tokens_per_batch=1000` to have a try? I have conducted experiments before and found that the limit batch tokens for 12GB gpu are nearly 1200.

**Comment by magician-david** [comment]
> @magician-david how many GPUs and data do you use?



> @magician-david can you set `tokens_per_batch=1000` to have a try? I have conducted experiments before and found that the limit batch tokens for 12GB gpu are nearly 1200.

For pre-training or fine-tuning? Or for both? Thanks!

**Comment by StillKeepTry** [comment]
@magician-david pre-training or fine-tuning will be ok.

**Comment by magician-david** [comment]
> @magician-david can you set `tokens_per_batch=1000` to have a try? I have conducted experiments before and found that the limit batch tokens for 12GB GPU are nearly 1200.
> 
> By the way, 1 GPU may really poor. To my experience, 8 GPUs can surpass 1 GPU nearly 7~8 points.

Thanks for your advice! I can use at most 4 GPUs. Maybe I should wait for the release of pre-trained model. When will the codes and models for other experiments release?

**Comment by StillKeepTry** [comment]
@magician-david We have released models for unsupervised nmt now. Here is [en-fr] model.(https://modelrelease.blob.core.windows.net/mass/mass_enfr_1024.pth).

**Comment by magician-david** [comment]
> @magician-david We have released models for unsupervised nmt now. Here is [en-fr](https://modelrelease.blob.core.windows.net/mass/mass_enfr_1024.pth) model.

Thanks for your prompt reply. I need a single English MASS model for language generation task.

**Comment by StillKeepTry** [comment]
> > @magician-david We have released models for unsupervised nmt now. Here is [en-fr](https://modelrelease.blob.core.windows.net/mass/mass_enfr_1024.pth) model.
> 
> Thanks for your prompt reply. I need a single English MASS model for language generation task.

We will release our english model and fine-tune code in a few days.

**Comment by magician-david** [comment]
> We will release our english model and fine-tune code in a few days.

Great! Thank you very much!
